compiler/types.nim

```diff
@@ -1185,173 +1185,377 @@ const
   szIllegalRecursion* = -2
   szUnknownSize* = -1
 
-proc computeSizeAux(typ: PType, a: var BiggestInt): BiggestInt
-proc computeRecSizeAux(n: PNode, a, currOffset: var BiggestInt): BiggestInt =
-  var maxAlign, maxSize, b, res: BiggestInt
+proc computeSizeAlign(typ: PType): void
+
+
+proc computeSubObjectAlign(n: PNode): BiggestInt =
+  ## returns object size and align
   case n.kind
   of nkRecCase:
     assert(n.sons[0].kind == nkSym)
-    result = computeRecSizeAux(n.sons[0], a, currOffset)
-    maxSize = 0
-    maxAlign = 1
-    for i in countup(1, sonsLen(n) - 1):
-      case n.sons[i].kind
+    result = computeSubObjectAlign(n.sons[0])
+
+    for i in 1 ..< sonsLen(n):
+      let child = n.sons[i]
+      case child.kind
       of nkOfBranch, nkElse:
-        res = computeRecSizeAux(lastSon(n.sons[i]), b, currOffset)
-        if res < 0: return res
-        maxSize = max(maxSize, res)
-        maxAlign = max(maxAlign, b)
-      else: internalError("computeRecSizeAux(record case branch)")
-    currOffset = align(currOffset, maxAlign) + maxSize
-    result = align(result, maxAlign) + maxSize
-    a = maxAlign
+        let align = computeSubObjectAlign(child.lastSon)
+        if align < 0:
+          return align
+        result = max(result, align)
+      else:
+        internalError("computeObjectOffsetsRecursive(record case branch)")
+
   of nkRecList:
-    result = 0
-    maxAlign = 1
-    for i in countup(0, sonsLen(n) - 1):
-      res = computeRecSizeAux(n.sons[i], b, currOffset)
-      if res < 0: return res
-      currOffset = align(currOffset, b) + res
-      result = align(result, b) + res
-      if b > maxAlign: maxAlign = b
-    a = maxAlign
+    result = 1
+
+    for i, child in n.sons:
+      let align = computeSubObjectAlign(n.sons[i])
+      if align < 0:
+        return align
+
+      result = max(result, align)
+
   of nkSym:
-    result = computeSizeAux(n.sym.typ, a)
-    n.sym.offset = int(currOffset)
+    n.sym.typ.computeSizeAlign
+    result = n.sym.typ.align
+
   else:
-    a = 1
-    result = szNonConcreteType
+    result = 1
+
+proc computeObjectOffsetsRecursive(n: PNode, initialOffset: BiggestInt): tuple[size, align: BiggestInt] =
```

Comment:
This is now enough code that it should reside it its own module. Suggested name: ``sizeofimpl.nim``